File: src/mixins/MaxSuccessiveWorkDays.js

/*
 * Mixin to compute fitness regarding maximum
 * successive work days of employees.
 */

'use strict';

const MaxSuccessiveWorkDays = (Base, props) => class extends Base {
  fitness() {
    let curr1
      , result = 0
      , employeesIter = props.employees();

    // Iterate through all employees
    while (curr1 = employeesIter.next()) {
      let curr2
        , successiveWorkDays = 0
        , periodDaysIter = props.periodDays();

      while (curr2 = periodDaysIter.next()) {
        if (this[props.shiftIndex(curr1, curr2)]) {
          successiveWorkDays += 1; // Day at work
        } else { // Free day
          // If maximum of successive work days is exceeded
          result += Math.min(
            props.opts.maxSuccessiveWorkDays - successiveWorkDays,
            0
          );

          // Reset the counter of successive work days
          successiveWorkDays = 0;
        }
      }

      // Check last days of period for successive work days
      result += Math.min(
        props.opts.maxSuccessiveWorkDays - successiveWorkDays,
        0
      );
    }

    if (__DEV__) {
      debugs.fitness.log(this, 'MaxSuccessiveWorkDays', result);
    }

    return super.fitness() + result;
  }
}

MaxSuccessiveWorkDays.requiredOpts = {
  maxSuccessiveWorkDays: true
};

module.exports = MaxSuccessiveWorkDays;

File: src/core/main/java/workschedule/schedule/fitnesses/MaxSuccessiveWorkDays.java
package workschedule.schedule.fitnesses;

import workschedule.Chromosome;
import workschedule.schedule.ScheduleDay;
import workschedule.schedule.ScheduleProperties;
import workschedule.schedule.options.MaxSuccessiveWorkDaysOption;

import java.util.Iterator;

public final class MaxSuccessiveWorkDays extends AbstractFitness {
    private final MaxSuccessiveWorkDaysOption option;

    public MaxSuccessiveWorkDays(final ScheduleProperties props) {
        super(props);
        option = opts.forClass(MaxSuccessiveWorkDaysOption.class);
    }

    @Override
    public int fitness(final Chromosome chr) {
        int result = 0;

        Iterator<Integer> employeeIterator = props.employees();

        // Iterate through all employees
        while (employeeIterator.hasNext()) {
            int successiveWorkDays = 0;

            Integer employee = employeeIterator.next();

            Iterator<ScheduleDay> scheduleDayIterator = props.scheduleDays();

            while (scheduleDayIterator.hasNext()) {
                ScheduleDay scheduleDay = scheduleDayIterator.next();

                if (chr.getParam(props.getShiftIndex(employee, scheduleDay))) {
                    successiveWorkDays++; // Day at work
                } else { // Free day
                    // If maximum successive work days is exceeded
                    result += Math.min(
                        option.get() - successiveWorkDays,
                        0
                    );

                    // Reset the counter of successive work days
                    successiveWorkDays = 0;
                }
            }

            // Check last days of schedule for successive work days
            result += Math.min(
                option.get() - successiveWorkDays,
                0
            );
        }

        return result;
    }
}
